src/main/java/io/airlift/compress/zstd/ZstdInputStream.java

```diff
@@ -0,0 +1,458 @@
+    }
+
+    int decodeRle()
+    {
+        check(inputAddress() + 1 <= inputLimit(), "Not enough input bytes");
+        check(outputAddress() + curBlockSize <= outputLimit(), "Not enough output space");
+        return blockDecompressor.decodeRleBlock(curBlockSize, inputBuffer, inputAddress(), outputBuffer, outputAddress(), outputLimit());
+    }
+
+    int decodeCompressed()
+    {
+        check(inputAddress() + curBlockSize <= inputLimit(), "Not enough input bytes");
+        check(outputAddress() + blockMaximumSize <= outputLimit(), "Not enough output space");
+        return blockDecompressor.decodeCompressedBlock(
+                inputBuffer, inputAddress(),
+                curBlockSize,
+                outputBuffer, outputAddress(), outputLimit(),
+                windowSize, ARRAY_BYTE_BASE_OFFSET);
+    }
+
+    private void decompressBlock()
+    {
+        check(outputPosition == outputEnd, "orphan output present");
+        switch (curBlockType) {
+            case RAW_BLOCK:
+                ensureOutputSpace(curBlockSize);
+                outputEnd += decodeRaw();
+                inputPosition += curBlockSize;
+                break;
+            case RLE_BLOCK:
+                ensureOutputSpace(curBlockSize);
+                outputEnd += decodeRle();
+                inputPosition += 1;
+                break;
+            case COMPRESSED_BLOCK:
+                check(curBlockSize < blockMaximumSize, "compressed block must be smaller than Block_Maximum_Size");
+                ensureOutputSpace(blockMaximumSize);
+                outputEnd += decodeCompressed();
+                inputPosition += curBlockSize;
+                break;
+            default:
+                throw fail(curInputFilePosition(), "Invalid block type " + curBlockType);
+        }
+        curBlockType = -1;
+        if (lastBlock) {
+            curHeader = null;
+            blockDecompressor = null;
+            if (contentChecksumFlag) {
+                check(inputAvailable() >= SIZE_OF_INT, "missing checksum data");
+                inputPosition += SIZE_OF_INT;
+            }
+        }
+    }
+
+    private int inputAvailable()
+    {
+        return inputEnd - inputPosition;
+    }
+
+    private int inputSpace()
+    {
+        return inputBuffer.length - inputEnd;
+    }
+
+    private long curInputFilePosition()
+    {
+        return evictedInput + inputPosition;
+    }
+
+    private void ensureInputSpace(int size)
+    {
+        if (inputSpace() < size) {
+            evictedInput += inputPosition;
+            if (size < inputPosition) {
+                System.arraycopy(inputBuffer, inputPosition, inputBuffer, 0, inputAvailable());
+                inputEnd = inputAvailable();
+                inputPosition = 0;
+            }
+            else {
+                int newSize = (inputBuffer.length * 2 + size) & BUFFER_SIZE_MASK;
```

Comment:
This will be `0` if the new size is otherwise too small, which can happen if the constructor specifying an input buffer size is used, with a small argument.  I also didn't check for the possibility of overflow.